### PR TITLE
Jakarta security Logout tests

### PR DESCRIPTION
**Pull request by c00crane**
Create Jakarta security logout tests.
Test logout functionality and the attributes of @LogoutDefinition.

### CODE REVIEW

**Comment by c00crane** [comment]
#build
#libby
#spawn.fullfat.buckets=io.openliberty.security.jakartasec.3.0.internal_fat,
io.openliberty.security.jakartasec.3.0.internal_fat.common,
io.openliberty.security.jakartasec.3.0.internal_fat.config,
io.openliberty.security.jakartasec.3.0.internal_fat.logout

**Comment by c00crane** [comment]
#build
#libby
#spawn.fullfat.buckets=io.openliberty.security.jakartasec.3.0.internal_fat,
io.openliberty.security.jakartasec.3.0.internal_fat.common,
io.openliberty.security.jakartasec.3.0.internal_fat.config,
io.openliberty.security.jakartasec.3.0.internal_fat.logout

**Comment by c00crane** [comment]
#libby
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_l7Js8F-HEe2zr6lGBCS8Nw

File: dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/expectations/ServerTraceExpectation.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
```
Comment (ayoho) [inline dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/expectations/ServerTraceExpectation.java:2]:
```suggestion
 * Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/utils/CommonExpectations.java [hunk 2]
```diff
@@ -110,4 +111,12 @@ public static Expectations successfullyReachedOidcLoginPage() {
         return expectations;
     }
 
+    public static Expectations successfullyReachedOidcLogoutPage() {
```
Comment (ayoho) [inline dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/utils/CommonExpectations.java:114]: I'd recommend putting this in a new, separate `com.ibm.ws.security.fat.common.oidc` project to make sure this project stays as truly common as possible.

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java [hunk 3]
```diff
@@ -300,6 +367,30 @@ public Expectations getProcessLoginExpectations(String app) throws Exception {
         return processLoginexpectations;
     }
 
+    public Expectations getProcessAppAccessWithRefreshtedTokenExpectations(String app) throws Exception {
```
Comment (ayoho) [inline dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java:370]:
```suggestion
    public Expectations getProcessAppAccessWithRefreshedTokenExpectations(String app) throws Exception {
```

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java [hunk 4]
```diff
@@ -451,4 +542,100 @@ public void validateTheSameContext(String instance, Page response) throws Except
...
+//        }
+    }
+
+    public boolean tokensAreTheSame(Page response1, Page response2, boolean provderAllowsRefresh) throws Exception {
+
+        if (provderAllowsRefresh) {
+            return !accessTokensAreDifferent(response1, response2) && !idTokensAreDifferent(response1, response2) && !refreshTokensAreDifferent(response1, response2);
+        } else {
+            return !accessTokensAreDifferent(response1, response2) && !idTokensAreDifferent(response1, response2);
+        }
```
Comment (ayoho) [inline dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java:578]:
```suggestion
        boolean tokensAreTheSame = !accessTokensAreDifferent(response1, response2) && !idTokensAreDifferent(response1, response2);
        if (provderAllowsRefresh) {
            return tokensAreTheSame && !refreshTokensAreDifferent(response1, response2);
        }
```

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonLogoutAndRefreshTests.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        Page response2 = invokeAppGetToAppWithRefreshedToken(webClient, url); // get to app not because either id or access token is good, but because the token was refreshed.
+
+        if (tokensAreDifferent(response1, response2, providerAllowsRefresh, TokenWasRefreshed)) {
+            Log.info(thisClass, _testName, "Test tokens were refreshed");
+        } else {
+            // TODO 23213
+            fail("Test tokens were NOT refreshed 23213");
+        }
+        Expectations expectations = new Expectations();
+        expectations.addExpectation(new ServerTraceExpectation(null, rpServer, Constants.STRING_DOES_NOT_MATCH, "Redirect to the OpenID Connect Provider Authentication endpoint for re-authentication", "The request did not result in a re-auth request."));
```
Comment (ayoho) [inline dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonLogoutAndRefreshTests.java:65]:
```suggestion
        expectations.addExpectation(new ServerTraceExpectation(null, rpServer, Constants.STRING_DOES_NOT_MATCH, "Redirect to the OpenID Connect Provider Authentication endpoint for re-authentication", "The request got redirected to the OP to re-authenticate when it shouldn't have."));
```

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/configs/TestConfigMaps.java [hunk 6]
```diff
@@ -119,7 +176,105 @@ public static Map<String, Object> getEmptyCallerGroupsClaim() throws Exception {
...
+    public static Map<String, Object> mergeMaps(Map<String, Object> updatedMap, Map<String, Object> newMap) throws Exception {
+
+        if (updatedMap == null) {
+            updatedMap = new HashMap<String, Object>();
+        }
+//        for (Map<String, Object> aMap : newMap.entrySet()) {
+        for (Entry<String, Object> x : newMap.entrySet()) {
+            updatedMap.put(x.getKey(), x.getValue());
+        }
+//        }
```
Comment (ayoho) [inline dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/configs/TestConfigMaps.java:277]:
```suggestion
        updatedMap.putAll(newMap);
```
I'm not sure this necessarily needs to be its own method, unless you just want one that will instantiate a new Map for you when `updatedMap` is null.

**Review by ayoho — APPROVED** [review]

File: dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/expectations/ServerTraceExpectation.java (hunk 1 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/expectations/ServerTraceExpectation.java:2]: updated - will deliver with refresh tests

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java (hunk 3 above)
Comment (c00crane) [reply dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java:370]: Don't know why I can't type the word without adding a "t" - I had already fixed many other instances before pushing the code.
updated - will deliver with refresh tests

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java (hunk 4 above)
Comment (c00crane) [reply dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonAnnotatedSecurityTests.java:578]: Updated both methods - will deliver with refresh tests

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonLogoutAndRefreshTests.java (hunk 5 above)
Comment (c00crane) [reply dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/commonTests/CommonLogoutAndRefreshTests.java:65]: updated - will deliver with refresh tests

File: dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/configs/TestConfigMaps.java (hunk 6 above)
Comment (c00crane) [reply dev/io.openliberty.security.jakartasec.3.0.internal_fat.common/src/io/openliberty/security/jakartasec/fat/configs/TestConfigMaps.java:277]: I wanted a method to merge a map into another without having to worry if the first map existed (not having to check for that in a ton of locations, ...)
I updated the method to use putAll - will deliver with refresh tests

File: dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/utils/CommonExpectations.java (hunk 2 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.fat.common/src/com/ibm/ws/security/fat/common/utils/CommonExpectations.java:114]: I moved it into the common project for the jakarta tests - if we see a need for it from other oidc projects, I'll create a new common project at that time.
update will be delivered with the refresh tests

**Pull request merged**